src/openeo_udf/api/datacube.py

```diff
@@ -277,6 +279,179 @@ def from_data_collection(data_collection: 'openeo_udf.server.data_model.data_col
         return dc_list
 
 
+    @staticmethod
+    def from_file(filename, fmt='netcdf') -> "DataCube":
+        if fmt.lower()=='netcdf':
+            return DataCube(DataCube._load_DataArray_from_NetCDF(filename))
+        if fmt.lower()=='json':
+            return DataCube(DataCube._load_DataArray_from_JSON(filename))
+    
+    
+    def to_file(self, filename, fmt='netcdf'):
+        if fmt.lower()=='netcdf':
+            return DataCube._save_DataArray_to_NetCDF(filename, self.get_array())
+        if fmt.lower()=='json':
+            return DataCube._save_DataArray_to_JSON(filename, self.get_array())
+    
+    
+    @staticmethod
+    def _load_DataArray_from_JSON(filename) -> xarray.DataArray:
+        with open(filename) as f:
+            # get the deserialized json
+            d=json.load(f)
+            d['data']=numpy.array(d['data'],dtype=numpy.dtype(d['attrs']['dtype']))
+            for k,v in d['coords'].items():
+                # prepare coordinate 
+                d['coords'][k]['data']=numpy.array(v['data'],dtype=v['attrs']['dtype'])
+                # remove dtype and shape, because that is included for helping the user
+                if d['coords'][k].get('attrs',None) is not None:
+                    d['coords'][k]['attrs'].pop('dtype',None)
+                    d['coords'][k]['attrs'].pop('shape',None)
+            
+            # remove dtype and shape, because that is included for helping the user
+            if d.get('attrs',None) is not None:
+                d['attrs'].pop('dtype',None)
+                d['attrs'].pop('shape',None)
+            # vonvert to xarray
+            r=xarray.DataArray.from_dict(d)
+            del d
+        # build dimension list in proper order
+        dims=list(filter(lambda i: i!='t' and i!='bands' and i!='x' and i!='y',r.dims))
+        if 't' in r.dims: dims+=['t']
+        if 'bands' in r.dims: dims+=['bands']
+        if 'x' in r.dims: dims+=['x']
+        if 'y' in r.dims: dims+=['y']
+        # return the resulting data array
+        return r.transpose(*dims)
+    
+    
+    @staticmethod
+    def _load_DataArray_from_NetCDF(filename) -> xarray.DataArray:
+        # load the dataset and convert to data array
+        ds=xarray.open_dataset(filename, engine='h5netcdf')
+        r=ds.to_array(dim='bands')
+        # build dimension list in proper order
+        dims=list(filter(lambda i: i!='t' and i!='bands' and i!='x' and i!='y',r.dims))
+        if 't' in r.dims: dims+=['t']
+        if 'bands' in r.dims: dims+=['bands']
+        if 'x' in r.dims: dims+=['x']
+        if 'y' in r.dims: dims+=['y']
+        # return the resulting data array
+        return r.transpose(*dims)
+    
+    
+    @staticmethod
+    def _save_DataArray_to_JSON(filename, array: xarray.DataArray):
+        # to deserialized json
+        jsonarray=array.to_dict()
```

Comment:
I would have expected you to use the DataCube.to_dict() method here, which is also intended to serialize cubes into json. Like this it seems we are introducing two different json formats for the same thing?

Reply:
With to_dict(), this for example fails, because numpy datetime64 is not serializable

import xarray
import numpy
from openeo_udf.api.datacube import DataCube
import json

if __name__ == '__main__':
    arr=xarray.DataArray(
        numpy.ones([2,2,2,2]),
        dims=['t','bands','x','y'],
        coords={
            't':[numpy.datetime64('2020-01-01'),numpy.datetime64('2020-01-02')],
            'bands':['b0','b1']
        }
    )
    td=json.dumps(DataCube(arr).to_dict())

Reply:
This might be VITO backend specific but this fails too:
import xarray
import numpy
from openeo_udf.api.datacube import DataCube
import json

if __name__ == '__main__':
    arr=xarray.DataArray(
        numpy.ones([2,2,2,2]),
        dims=['t','bands','x','y'],
        coords={
            't':[numpy.datetime64('2020-01-01'),numpy.datetime64('2020-01-02')],
            'bands':['b0','b1']
        }
    )
    DataCube.from_dict(DataCube(arr).to_dict())

This is the form how a DataCube is given to an apply_dimension.

The to_dict method loses the connection to which axis a coordinate belongs to if not all dims have coords. I know there is the convention ('t','bands','x','y'), but that will always cause rework when new possible variables added. IMHO adding the dimension list into the dict could make things more robust.